### PR TITLE
Disable cleanup of test cases once they have been run

### PR DESCRIPTION
**Pull request by cottsay**
From the [`unittest` docs](https://docs.python.org/3/library/unittest.html#unittest.TestSuite):
> After TestSuite.run(), callers should not rely on the tests returned by this method unless the caller uses a subclass that overrides TestSuite._removeTestAtIndex() to preserve test references.

If the test run is interrupted, `_LaunchDiedException` triggers a `FailResult` for the entire run. However, some cases may have already been cleaned up by the test run, and `FailResult` calls into `unittest.TestResult.addFailure()` with possibly cleaned-up tests, which are represented by `None`.

This leads to traces like this:
```
Traceback (most recent call last):
  File "/opt/ros/master/bin/ros2", line 11, in <module>
    load_entry_point('ros2cli', 'console_scripts', 'ros2')()
  File "/opt/ros_src/master/build/ros2cli/ros2cli/cli.py", line 67, in main
    rc = extension.main(parser=parser, args=args)
  File "/opt/ros_src/master/build/ros2test/ros2test/command/test.py", line 48, in main
    parser, args, test_runner_cls=launch_testing_ros.LaunchTestRunner
  File "/opt/ros_src/master/build/launch_testing/launch_testing/launch_test.py", line 111, in run
    results = runner.run()
  File "/opt/ros_src/master/build/launch_testing/launch_testing/test_runner.py", line 267, in run
    message='Launch stopped before the active tests finished.'
  File "/opt/ros_src/master/build/launch_testing/launch_testing/test_result.py", line 26, in __init__
    self.addFailure(case, (Exception, Exception(message), None))
  File "/usr/lib64/python3.7/unittest/result.py", line 17, in inner
    return method(self, *args, **kw)
  File "/usr/lib64/python3.7/unittest/result.py", line 122, in addFailure
    self.failures.append((test, self._exc_info_to_string(err, test)))
  File "/usr/lib64/python3.7/unittest/result.py", line 180, in _exc_info_to_string
    if exctype is test.failureException:
AttributeError: 'NoneType' object has no attribute 'failureException'
```

The solution presented here is to change the `_removeTestAtIndex` function to a no-op, since we aren't actually subclassing `unittest.TestSuite` as the documentation suggests.

### CODE REVIEW

File: launch_testing/launch_testing/loader.py [hunk 1]
```diff
@@ -285,6 +292,17 @@ def _iterate_test_classes_in_test_suite(test_suite):
             yield t.__class__
 
 
+def _iterate_test_suites(test_suite):
+    try:
+        iter(test_suite)
+    except TypeError:
```
Comment (ivanpauno) [inline launch_testing/launch_testing/loader.py:298]: what is this catching? Shouldn't `pre_shutdown_tests`/`post_shutdown_tests` always be an iterable?

Comment (cottsay) [reply launch_testing/launch_testing/loader.py:298]: The `test_suite` object is not iterable if it is actually a test instance and not a test suite containing a test instances.

This is the same logic that is used in the enumeration of only the test suites (the opposite of this new function): https://github.com/ros2/launch/blob/d16628d8abfcdd7d44293c89198a1a64cfc54ee6/launch_testing/launch_testing/loader.py#L288-L293

**Comment by cottsay** [comment]
> How the original error can be reproduced?

Hit Ctrl-C while running a launch test where at least one of the tests has already completed. For example:
<details>

```
$ ros2 test src/ros2/launch/launch_testing/test/launch_testing/examples/terminating_proc_launch_test.py 
Running with ROS_DOMAIN_ID 1
ROS_DOMAIN_ID 1
Warning: Passing ready_fn as an argument to generate_test_description will be removed in a future release.  Include a launch_testing.actions.ReadyToTest action in the LaunchDescription instead.
[INFO] [launch]: All log files can be found below /home/cottsay/.ros/log/2020-05-11-11-52-24-276581-cottsay-lenovo.delphi.cottsay.net-14222
[INFO] [launch]: Default logging verbosity is set to INFO
test_no_args (terminating_proc_launch_test.TestTerminatingProc)
Test terminating_proc without command line arguments. ... [INFO] [terminating_proc-1]: process started with pid [14225]
[terminating_proc-1] Starting Up
[terminating_proc-1] Ready
[terminating_proc-1] Emulating Work
[terminating_proc-1] Done
[terminating_proc-1] Shutting Down
[INFO] [terminating_proc-1]: process has finished cleanly [pid 14225]
ok
test_unhandled_exception (terminating_proc_launch_test.TestTerminatingProc)
Test terminating_proc forcing an unhandled exception. ... [INFO] [terminating_proc-2]: process started with pid [14227]
[terminating_proc-2] Starting Up
^C[WARNING] [launch]: user interrupted with ctrl-c (SIGINT)
[terminating_proc-2] Traceback (most recent call last):
[terminating_proc-2]   File "/opt/ros/master/lib/launch_testing/terminating_proc", line 26, in <module>
[terminating_proc-2]     time.sleep(1.0)
[terminating_proc-2] KeyboardInterrupt
[ERROR] [terminating_proc-2]: process has died [pid 14227, exit code -2, cmd '/usr/bin/python3 /opt/ros/master/lib/launch_testing/terminating_proc --exception'].
Processes under test stopped before tests completed
Process 'terminating_proc-2' exited with -2
##### 'terminating_proc-2' output #####
Starting Up

Traceback (most recent call last):
  File "/opt/ros/master/lib/launch_testing/terminating_proc", line 26, in <module>
    time.sleep(1.0)
KeyboardInterrupt

#######################################
Traceback (most recent call last):
  File "build/launch_testing/launch_testing/test_runner.py", line 261, in run
    results[run] = worker.run()
  File "build/launch_testing/launch_testing/test_runner.py", line 175, in run
    raise _LaunchDiedException()
launch_testing.test_runner._LaunchDiedException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ros/master/bin/ros2", line 11, in <module>
    load_entry_point('ros2cli', 'console_scripts', 'ros2')()
  File "build/ros2cli/ros2cli/cli.py", line 67, in main
    rc = extension.main(parser=parser, args=args)
  File "build/ros2test/ros2test/command/test.py", line 47, in main
    return launch_testing.launch_test.run(
  File "build/launch_testing/launch_testing/launch_test.py", line 111, in run
    results = runner.run()
  File "build/launch_testing/launch_testing/test_runner.py", line 270, in run
    results[run] = FailResult(
  File "build/launch_testing/launch_testing/test_result.py", line 26, in __init__
    self.addFailure(case, (Exception, Exception(message), None))
  File "/usr/lib64/python3.8/unittest/result.py", line 17, in inner
    return method(self, *args, **kw)
  File "/usr/lib64/python3.8/unittest/result.py", line 122, in addFailure
    self.failures.append((test, self._exc_info_to_string(err, test)))
  File "/usr/lib64/python3.8/unittest/result.py", line 180, in _exc_info_to_string
    if exctype is test.failureException:
AttributeError: 'NoneType' object has no attribute 'failureException'
```
</details>

> I'm not sure how this PR avoids the mentioned error.

The error occurs because `FailResult` is assuming that all of the tests cases can be used when calling `addFailure` in the creation of a `TestResult` instance: https://github.com/ros2/launch/blob/d16628d8abfcdd7d44293c89198a1a64cfc54ee6/launch_testing/launch_testing/test_result.py#L23-L27

In the current implementation, test cases that have already completed have been cleaned up and can not be passed to `addFailure`.

If we just skip `addFailure` for the test cases that have already been completed, we'll lose them entirely, since the current structure of this code doesn't have the capability to report partial success if launch shuts down before all of the tests finish.

File: launch_testing/launch_testing/loader.py (hunk 1 above)
Comment (cottsay) [reply launch_testing/launch_testing/loader.py:298]: Ah, the CI has demonstrated that the inverse of the logic in `_iterate_tests_in_test_suite` doesn't hold true here. I'll have to find a better way to enumerate the suites.

**Comment by hidmic** [comment]
@cottsay ping.

**Comment by cottsay** [comment]
> Since we're monkey patching either way, can't we set [`TestSuite._cleanup`](https://github.com/python/cpython/blob/a63c61168588937c482435d0432c753de4844c46/Lib/unittest/suite.py#L19) to `False` instead?

I'd like to use `_removeTestAtIndex` because it is the documented method for falling back to the previous behavior of not cleaning up the cases after success. Because it is documented like this, I believe it is unlikely to change and essentially part of the API, whereas `_cleanup` isn't mentioned and seems more like a typical private member (which would really be "monkey patching".

> ...Can't we use [`addSuccess()`](https://docs.python.org/3/library/unittest.html#unittest.TestResult.addSuccess) and [`addSkip()`](https://docs.python.org/3/library/unittest.html#unittest.TestResult.addSkip) appropriately instead? 

At the point where we're currently calling `addFailure`, the case is already `None` because of the cleanup. Even after suppressing the cleanup, the `TestCase` doesn't store the result so we don't know which tests were successful, failed OR ignored - that is what the `TestResult` class is for. It might be possible to get access to the `TestResult` instance that has the accumulated successful and ignored cases, but I can't see a straightforward way to get access to it. It might not even be possible.

**Comment by hidmic** [comment]
> I'd like to use _removeTestAtIndex because it is the documented method

Fair enough.

> At the point where we're currently calling addFailure, the case is already None because of the cleanup. 

Correct, you'd have to propagate it upwards, perhaps along with the `LaunchDiedException`. At any rate, it's a nice to have, if launch dies the test is broken anyways.

File: launch_testing/launch_testing/loader.py (hunk 1 above)
Comment (cottsay) [reply launch_testing/launch_testing/loader.py:298]: Resolved by b604a34

**Pull request merged**